### PR TITLE
Trig functions and NCL receipt

### PR DESCRIPTION
**Pull request by cyschneck**
## PR Summary
Trig functions

- sin
- cosine
- tangent
- arcsine
- arccosine
- arctangent

## PR Checklist
**General**
- [x] PR includes a summary of changes
- [x] Link relevant issues, make one if none exist

### CODE REVIEW

**Comment by jukent** [comment]
Looks great. Thanks for the updates to the Contributing guide as well. It seems to be missing from the homepage though?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -3,24 +3,25 @@
 ## Create the docs locally
 
 ``` python
-conda env update -f environment.yml
+conda env create --file environment.yml
 conda activate geocat-applications
 make clean html
 open _build/html/index.html
 ```
+If the `_build` folder is generated, but `open` command cannot open the html, you can navigate to the `index.html` document and right-click to `Open With` your desired browser.
```
Comment (anissa111) [inline CONTRIBUTING.md:11]: I wonder if this is a mac thing because `open _build/html/index.html` works as expected for me

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -104,3 +105,7 @@ than waiting for a full **NCL Entry** to be completed.
 ### Adding to the NCL Index
 
 Add a new line to the `ncl/ncl_index/ncl-index-table.csv` file
+
+| NCL Function       | Description                 | Python Equivalent | Notes                          |
+|--------------------|-----------------------------|-------------------|--------------------------------|
+| `example_function <https://www.ncl.ucar.edu/>`__ | Description of the function | `python_function` | Link to `ncl_entries` notebook |
```
Comment (anissa111) [inline CONTRIBUTING.md:111]: Is this going to be confusing because of markdown vs csv formatting?

**Comment by anissa111** [comment]
Just a few initial thoughts, I'll give this a more thorough look tomorrow 👍

File: CONTRIBUTING.md (hunk 1 above)
Comment (cyschneck) [reply CONTRIBUTING.md:11]: It might be, it gave me issues running on Linux, even though the html was being generated

Comment (anissa111) [reply CONTRIBUTING.md:11]: Hmm, yeah. How do you feel about leaving it like this here and making a separate issue to expand this section into a proper "Development workflow" section (like in [geocat-comp](https://geocat-comp.readthedocs.io/en/stable/contrib.html#development-workflow-overview) or [xarray](https://docs.xarray.dev/en/stable/contributing.html#development-workflow))?

File: CONTRIBUTING.md (hunk 2 above)
Comment (anissa111) [reply CONTRIBUTING.md:111]: Yeah I found it really clunky to edit, too, which might be problematic for outside contributors. I wonder if a yaml or json file or some other format would be easier. Even if we couldn't directly transform whatever file type into a table in the rst file, we could generate a csv in `conf.py`.

@kafitzgerald I remember you had some thoughts about this previously, do you have any opinions here?

File: CONTRIBUTING.md (hunk 1 above)
Comment (cyschneck) [reply CONTRIBUTING.md:11]: Sounds good, will do

File: ncl/ncl_entries/sin_trigonometry.ipynb [hunk 3]
```diff
@@ -0,0 +1,282 @@
+{
```
Comment (anissa111) [inline ncl/ncl_entries/sin_trigonometry.ipynb:1]: <p>This might be picky of me, but I think we should call this page either Sine or Trigonometry (with the assumption that we'd expand the section later or add a WIP admonition or something if we wanted to publish with just the sin functions).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (anissa111) [inline ncl/ncl_entries/sin_trigonometry.ipynb:1]: <p>I think we can include actual math formatting here, I think that may work better than a code block.</p><p> </p><p>Although, I think it might be okay to assume that we don't need to explain the math behind sine?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (anissa111) [inline ncl/ncl_entries/sin_trigonometry.ipynb:1]: <p>Might not be relevant with other comments, but I'd personally recommend numpy's sine function over the math library, especially since it can handle arrays</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (anissa111) [inline ncl/ncl_entries/sin_trigonometry.ipynb:1]: <p>Why is this part here at the bottom?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

**Comment by anissa111** [comment]
Sorry, the review notebook app didn't let me leave a top-level review comment.

I'm thinking we should either refocus this notebook to be a receipt, or at least a trig focused ncl entry if there's anything that a receipt notebook produces that's significant enough to warrant adding an ncl notebook.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -3,24 +3,24 @@
 ## Create the docs locally
 
 ``` python
-conda env update -f environment.yml
+conda env create --file environment.yml
 conda activate geocat-applications
 make clean html
 open _build/html/index.html
```
Comment (kafitzgerald) [inline CONTRIBUTING.md:9]: Again, maybe this is or should be handled elsewhere, but I like the language and suggestion from xarray if we're looking for something less macOS specific: https://docs.xarray.dev/en/stable/contributing.html#building-the-documentation

File: CONTRIBUTING.md (hunk 2 above)
Comment (kafitzgerald) [reply CONTRIBUTING.md:110]: Yeah, I'd maybe be inclined to just point to the CSV from here and describe what goes there are a high level if needed or any special cases (these can come later to if/when we feel pain).  It's a bummer that it was kind of a pain / confusing the first time around, but I think once that's populated a bit it should be more obvious and/or we'll need to write different instructions anyway 🙃. 

It's nice to have a visual, but I don't know that it's worth spending time on getting something in here (and maintaining it) that's similar enough not to be confusing.  

Just my opinion though.

File: ncl/ncl_entries/sin_trigonometry.ipynb (hunk 3 above)
Comment (kafitzgerald) [reply ncl/ncl_entries/sin_trigonometry.ipynb:1]: Yeah, this is probably more of a design / discussion point, but I think we may want to scale up the category here to at least trigonometry if not something like mathematical functions and keep it pretty light on description.

Comment (kafitzgerald) [reply ncl/ncl_entries/sin_trigonometry.ipynb:1]: Yeah, this is a personal take, but I think some of the relevant content for trig functions in Python that I'd want to know is more along the lines of where the find the functions and what ones to use when (probably numpy ones most often for performance reasons and compatibility w/ xarray, dask, and the "Pangeo" stack).  

Again this is totally a design thing.

**Review by kafitzgerald — COMMENTED** [review]
This seems to fit the rough template here and overall looks good, but I wonder if functions like this should be more of a special case where for the NCL entries portion where we have more of a general category and discuss differences at that level rather than individually for a few reasons:
* To keep the number of categories/items in the NCL entries bit more reasonable from a visual perspective (if we have a page that displays these)
* To focus more on the level at which there are commonalities/differences so it's easier to digest for any readers
* To limit the number of these we write and conserve some dev time

Happy to chat more about this and I'd love to hear other perspectives.

**Comment by cyschneck** [comment]
Ok, based on suggestions, I expanded this to cover the trig functions (sin, cos, tan, asin, acos, atan) since they are all very similar in execution. The notebook is now broken into:
- Python package options
- "Grab and Go" sections for each function (with hyperlinks)
- Overview of differences from NCL
- Additional Python resources

File: CONTRIBUTING.md (hunk 2 above)
Comment (cyschneck) [reply CONTRIBUTING.md:110]: I will add details about how to do this in the CONTRIBUTING.md (as part of #50) for now

File: ncl/ncl_entries/sin_trigonometry.ipynb (hunk 3 above)
Comment (cyschneck) [reply ncl/ncl_entries/sin_trigonometry.ipynb]: Since they are all using numpy/math I combined them into the most recent notebook with an overview at the top

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -3,24 +3,24 @@
 ## Create the docs locally
 
 ``` python
-conda env update -f environment.yml
+conda env create --file environment.yml
```
[the remark this answers is not in the document]
Comment (cyschneck) [reply CONTRIBUTING.md:6]: Either way works on my end as well, but the `create` option is clearer about what the script is doing, since `update` implies it is modifying something that is already existing which can be confusing is someone is less familiar with conda

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -3,24 +3,24 @@
...
 ## Types of Content
 
-There are several ways to contribute content to this repository.
+There are several ways to contribute content to this repository
 
 ### Python Entry
 
 Python entries are the content on the main page of the website. These entries
-are python-first content that do not require any knowledge of or reference NCL.
+are Python-first content that do not require any knowledge of or reference NCL.
```
[the remark this answers is not in the document]
Comment (cyschneck) [reply CONTRIBUTING.md:19]: I'll add it to #50

File: CONTRIBUTING.md (hunk 4 above)
Comment (cyschneck) [reply CONTRIBUTING.md:9]: I'll add it to #50

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kafitzgerald) [reply CONTRIBUTING.md:6]: Yeah, that's fair.  Sometimes the messages when you already have the environment are a bit confusing as well, but I think this is fine.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -30,15 +30,15 @@ NCL-specific content.*
 ### NCL Entry
```
[the remark this answers is not in the document]
Comment (kafitzgerald) [reply CONTRIBUTING.md:30]: I brought this up elsewhere. Resolving this comment.

File: ncl/ncl_entries/trigonometric_functions.ipynb [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+    "\n",
+    "The [`math`](https://docs.python.org/3/library/math.html#trigonometric-functions) and [`cmath`](https://docs.python.org/3/library/cmath.html#trigonometric-functions) module are part of the Python Standard Library and require no additional downloading to use. Both math modules can be used to calculate the value of trigonometric functions in Python\n",
+    "\n",
+    "Both math modules expect an input value in radians:\n",
+    "- Input: single input values (float, decimals, integars)\n",
+    "- Returns: single numerical value in radians\n",
+    "\n",
+    "The `math` function only work on real number inputs, while `cmath` contains equivalent functions for for complex and imaginary numbers\n",
+    "\n",
+    "However, if you need to work with lists or arrays neither math module will not be sufficient"
```
Comment (kafitzgerald) [inline ncl/ncl_entries/trigonometric_functions.ipynb:45]:
```suggestion
    "However, if you are working with lists or arrays we suggest using the numpy equivalents"
```

Given that you can work around the list/array limitation this seems like it might skip that rabbit hole and skip to the point.  We can certainly add more detail as well, but that seems like a later problem or if someone asks.

File: ncl/ncl_entries/trigonometric_functions.ipynb [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "42cd2476-00a3-4f29-a8be-ab19af9f5c69",
+   "metadata": {},
+   "source": [
+    "## Recommended: Using `numpy`\n",
+    "\n",
+    "[`numpy`](https://numpy.org/doc/stable/reference/routines.math.html) supports both real and complexvalues of individual numbers (floats, decimals, integars) as well as array-like values (arrays, lists, etc...)\n",
```
Comment (kafitzgerald) [inline ncl/ncl_entries/trigonometric_functions.ipynb:55]:
```suggestion
    "[`numpy`](https://numpy.org/doc/stable/reference/routines.math.html) supports both real and complex values of individual numbers (floats, decimals, and integers) as well as array-like values (arrays, lists, etc...)\n",
```

File: ncl/ncl_entries/trigonometric_functions.ipynb [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+    "\n",
+    "`numpy` expects an input value in radians:\n",
+    "- Input: array_list values (floats, integars, arrays, lists, etc..)\n",
```
Comment (kafitzgerald) [inline ncl/ncl_entries/trigonometric_functions.ipynb:58]: maybe something like "array, list, or similar of a numeric type"

File: ncl/ncl_entries/trigonometric_functions.ipynb [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "4a671627-ddc3-4936-b8a7-9e00ec821a24",
+   "metadata": {},
+   "source": [
+    "### Input Type\n",
+    "While the math modules (`math` and `cmath`) only accept single numerical values—like NCL—`numpy` can take in either single numerical values or array-like values. Unlike NCL, `cmath` and `numpy` can also accept complex numbers as input values\n",
+    "\n",
+    "### Numerical Differences\n",
+    "Both `math/cmath` and `numpy` return values with more points of precision than NCL. For example, in NCL `sin(0.5) = 0.4794255`, while in Python:"
```
Comment (kafitzgerald) [inline ncl/ncl_entries/trigonometric_functions.ipynb:595]: I think this might be from printing.  NCL should return either a float or a double depending upon the input value.

We might be able to leave this comment out and just talk about input and return types (if that's even needed - there's already some guidance above, but maybe folks have thoughts on that).

**Review by kafitzgerald — CHANGES_REQUESTED** [review]
I think this looks a lot better to from the design perspective (sorry you had to restructure this and come along for that ride).  

Mostly some relatively minor comments / suggestions here.

**Comment by cyschneck** [comment]
I removed the changes made to the contributor's guide and put them into the relevant PR #52 to clean up this PR

**Review by kafitzgerald — APPROVED** [review]
Looks good to me!  

Sorry there was so much back and for on design / structure.  I don't think any of us realized how many considerations weren't really well templated / documented / clear.

I think having this in there will be really helpful!

**Review by anissa111 — CHANGES_REQUESTED** [review]
I think I'll need to make another pass once the merge conflicts are incorporated, but two things I noticed:
- the trig notebook should link to the ncl docs for each function covered
- the ncl functions do accept multidimensional input (see https://www.ncl.ucar.edu/Document/Functions/Built-in/sin.shtml)

I'm slightly concerned that skipping "receipts" (sorry I know I know), even for super simple functions could lead to "misses" like the multidimensional bit here. I really think for everything we add to the index, we should actually run some ncl code, just to make totally sure we've done our due diligence.

File: ncl/ncl_entries/trigonometric_functions.ipynb [hunk 12]
```diff
@@ -0,0 +1,633 @@
+{
```
Comment (anissa111) [inline ncl/ncl_entries/trigonometric_functions.ipynb:1]: <p>Coming back in with a few more thoughts I had last night.</p><p> </p><p>I think that it would make more sense to link to the ncl documentation here instead of the section heads, especially since sphinx is providing us with a floating section navigator bit on the right.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (anissa111) [inline ncl/ncl_entries/trigonometric_functions.ipynb:1]: <p>I don't think we need to link every single reference of the modules, but I feel like we should link to the docs for <code>math</code>, <code>cmath</code>, and <code>numpy</code> and in the <code>numpy</code>  recommendation section.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (anissa111) [inline ncl/ncl_entries/trigonometric_functions.ipynb:1]: <p>Probably a broader discussion, but part of the "grab and go" appeal was that it was right up top and was our "recommendation". With having multiple sections with multiple options each, I think we're missing that.</p><p> </p><p>What would folks think about doing something this just once at the top?</p><p> </p><h2>Grab and Go</h2>

<pre class="ql-syntax" spellcheck="false">import numpy as np

rads = [0, np.pi/2, np.pi, 3*np.pi/2, 2*np.pi]

s = np.sin(rads)
c = np.cos(rads)
t = np.tan(rads)
as = np.asin(rads)
...

print(f"sin: {s}")
...
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (cyschneck) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: I have it split up since on the NCL index page they are separate functions and having separate sections lets the index hyperlink to the section of the notebook that is relevant

Comment (anissa111) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: I suppose I can see your point there. I'm curious what other folks think.

@jukent, I believe the grab and go sections were your idea originally, do you have any thoughts? Also @kafitzgerald, any opinions?

In a case where we might end up having different grab and go sections (here or elsewhere), I still think we should lean towards only having one recommendation in the grab and go unless there's a specific reason not to (like, use `x` if you're doing `y` workflow, `b` works better with `c` use cases).

Comment (kafitzgerald) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: I'm a little split between a) merge as-is and see how that looks and feels along with everything else b) an inclination to try and keep this a bit more simple and aligned with for instance the days_in_month NCL section grab and go.

I do think it's maybe a bit verbose as is and we may want to reorder the structure at some point, but I also don't know that we have a super clear idea of what will work best at this point and I think that might be ok.  There could be some value in just getting a few examples in and then re-evaluating what will work best. 
 Especially this early on I don't know that we can expect to get everything "right" and templated out in a way that is robust.

That said, it can be hard to come back and refactor / rearrange, but if we're intentional about it - I'm not too worried.

Comment (anissa111) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: Okay, I'll drop my concern for the purpose of merging this PR, we can talk about this particular point in a broader conversation.

**Comment by cyschneck** [comment]
Since the trig functions are using the same notebook, I don't think it will be able to populated by NCL Applications side bar since that is populated by separate notebooks

File: ncl/ncl_entries/trigonometric_functions.ipynb (hunk 12 above)
Comment (kafitzgerald) [inline ncl/ncl_entries/trigonometric_functions.ipynb:1]: <p>Does this need an update?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: ncl/receipts/trigonometric_functions.ipynb [hunk 13]
```diff
@@ -0,0 +1,275 @@
+{
```
Comment (kafitzgerald) [inline ncl/receipts/trigonometric_functions.ipynb:1]: <p>Might want to remove this before we merge.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (kafitzgerald) [inline ncl/receipts/trigonometric_functions.ipynb:1]: <p>Since this is the NCL receipt, should these be the NCL function names?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (kafitzgerald) [inline ncl/receipts/trigonometric_functions.ipynb:1]: <p>Are these diffs real?  If you provide a double to NCL's sin you should return a double as well which I imagine would be closer.  </p><p> </p><p>I think Python's built in float type that Math uses is double precision. </p><p> </p><p>Might be good to have some discussion of this somewhere (if it's not already elsewhere)</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

File: ncl/ncl_entries/trigonometric_functions.ipynb (hunk 12 above)
Comment (kafitzgerald) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: Nevermind, I see what's going on now.

File: ncl/receipts/trigonometric_functions.ipynb (hunk 13 above)
Comment (cyschneck) [reply ncl/receipts/trigonometric_functions.ipynb:1]: NCL is using a `typeof(0.5) = float` and Python is using `type(0.5) = float`. Am I missing a NCL feature? When a value is converted to a double, it still returns the same level of precision when using `math`
```
math.asin(np.double(0.5))
0.5235987755982989
```

Comment (kafitzgerald) [reply ncl/receipts/trigonometric_functions.ipynb:1]: There are some diffs in the definitions depending upon the context.  

In NCL the `float` type is 32 bits and `double` 64 bits, whereas in Python the `float` type is generally a 64-bit floating point number.  

This is also slightly different from numpy which has its own types (e.g. `np.float64`, `np.float32`).  

Generally the 64-bit types are the defaults in Python though, which is different from NCL and something that's come up in the past.

It seems like this is also what's causing the diffs here.

File: ncl/ncl_entries/trigonometric_functions.ipynb (hunk 12 above)
Comment (cyschneck) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: Do you mean that the references/links for the docs should be moved to the last section? I am currently linking them the first time I reference them and then again at end in Python Resources

File: ncl/receipts/trigonometric_functions.ipynb (hunk 13 above)
Comment (cyschneck) [reply ncl/receipts/trigonometric_functions.ipynb:1]: I can expand on the reasons behind the differences in this section of the receipt, but should it be rewritten to match? Currently, the `assert` statement is written to match within 6 decimal points
```
assert math.isclose(
        ncl_results[c], results_math[c], rel_tol=1e-06
    )  # within 6 decimal points
```

Comment (kafitzgerald) [reply ncl/receipts/trigonometric_functions.ipynb:1]: Maybe depends on what we're looking for in the comparison.  If we're not looking for precision level diffs (@anissa111 thoughts on this? -  you probably have a better idea how often this comes up or not), I think it's fine as is with a note somewhere in here about why.

Comment (anissa111) [reply ncl/receipts/trigonometric_functions.ipynb:1]: My impression is that this is going to come up a lot. I think it would make sense to make a full entry about the precision differences in the future and link that where/when it becomes relevant.

File: ncl/ncl_entries/trigonometric_functions.ipynb (hunk 12 above)
Comment (anissa111) [reply ncl/ncl_entries/trigonometric_functions.ipynb:1]: <p>Oh my bad, sorry, I missed the links on the code formatted text on reviewnb, never mind!</p>

**Review by anissa111 — APPROVED** [review]
